### PR TITLE
enable write_data for atom styles with bonus data

### LINKED ISSUE #1327: [BUG] Incorrect handling of atom_style ellipsoid data with read_data and write_data
**Summary**

Using `read_data` and `write_data` does not enforce or create correct data files

**LAMMPS Version and Platform**

LAMMPS 4 Jan 2019 and 8 Feb 2018, Linux x86_64 fedora 29 and ubuntu
 
**Expected Behavior**

`read_data` should enforce consistent settings in the data file, and `write_data` should write out correct and consistent data files.

**Actual Behavior**

`read_data` accepts lines in the `Atoms` section that set the ellipsoid flag, even though there are no ellipsoids defined in the header and thus no bonus storage allocated. Consequently, also the necessary `Ellipsoids` section is not required. `write_data` does not write out this information, thus the data file can be read in, but will cause segfaults due to the missing bonus data.
_Describe the actual behavior, how it differs from the expected behavior, and how this can be observed.  Try to be specific and do **not** use vague terms like "doesn't work" or "wrong result".  Do not assume that the person reading this has any experience with or knowledge of your specific area of research._

**Steps to Reproduce**

running `lmp -in in.ellipse.txt` will cause segfaults when trying to change the ellipse parameters or starting a run. The data file is inconsistent and `read_data` does not catch it.

Adding a `write_data` command to the gay-berne ellipsoid example produces an incorrect data file.

**Further Information, Files, and Links**

reported on [lammps-users](https://sourceforge.net/p/lammps/mailman/message/36585849/)

[data.txt](https://github.com/lammps/lammps/files/2859408/data.txt)
[in.ellipse.txt](https://github.com/lammps/lammps/files/2859409/in.ellipse.txt)

### PR DESCRIPTION
**Pull request by sjplimp**
**Summary**

Enable atoms styles with bonus data, including hybrid atom styles to write their bonus data to data files via the write_data command.

**Related Issues**

Closes #1327 

**Author(s)**

Steve Plimpton, SNL and Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.

### CODE REVIEW

**Comment by sjplimp** [comment]
@akohlmey This needs testing.  Hybrid should work (for combo of ellipsoid, try, line).  But body does not yet work.
Also not clear if the write_data_bonus() methods will output full precision, which they should for most of the floating point fields.

**Comment by akohlmey** [comment]
> @akohlmey This needs testing. Hybrid should work (for combo of ellipsoid, try, line). But body does not yet work.
> Also not clear if the write_data_bonus() methods will output full precision, which they should for most of the floating point fields.

will look into it. I already have a testing tool for a few atom styles, so it should be straightforward to add tests to it for ellipsoid and others.

**Comment by akohlmey** [comment]
@sjplimp this has a first round of tests and bugfixes. The code can now write and read back a data file for an all-ellipsoid system and have the same data. more later.

File: src/replicate.cpp [hunk 1]
```diff
@@ -188,6 +188,11 @@ void Replicate::command(int narg, char **arg)
   atom->ndihedrals = old->ndihedrals * nrep;
   atom->nimpropers = old->nimpropers * nrep;
 
+  atom->nellipsoids = old->nellipsoids * nrep;
```
Comment (sjplimp) [inline src/replicate.cpp:191]: good catch here

**Comment by akohlmey** [comment]
@ohenrich FYI the changes in this branch should fully enable writing and reading data and binary restart files for your models. I am in the process of completing tests, but atom style hybrid bond ellipsoid and atom style hybrid full ellipsoid have been successfully tested. same for using the replicate and reset_atom_ids (previously reset_ids) commands

**Comment by ohenrich** [comment]
Excellent! Many thanks for letting me know and for implementing this. We will test this soon in practice and put our old workaround aside.

**Comment by sjplimp** [comment]
@akohlmey Source code for all bonus styles is now there.  I'll be surprised if it works the first time.  Howto_body doc page still likely still needs some updating.

**Comment by akohlmey** [comment]
status update: as of today, also atom style "tri" is tested successfully and a few issues in its implementation were fixed in the process. this leaves the body styles... i had to merge in the `reset_mol_ids` changes, since some of the bugfixes there affect the tests here (applying `reset_atom_ids` is part of the testing process).

**Comment by sjplimp** [comment]
@akohlmey I finished the documention for this with input from Trung.  So once it is produces correct data files for all the new cases now supported, it is good to merge.  I suggest @ndtrung81 as one of the reviewers, particularly for the Howto_body doc page.

File: doc/src/Howto_body.rst [hunk 2]
```diff
@@ -219,21 +201,25 @@ in the *Bodies* section of the data file:
...
+each vertex. The diameter value can be different for each body
+particle. These floating-point values can be listed on as many lines
+as you wish; see the :doc:`read_data <read_data>` command for more
+details.
+
+.. note::
+
+  It is important that the vertices for each polygonal body particle be
+  listed in order around its perimeter, so that edges can be inferred.
+  LAMMPS does not check that this is the case.
```
Comment (ndtrung81) [inline doc/src/Howto_body.rst:223]: looks good for me

File: doc/src/Howto_body.rst [hunk 3]
```diff
@@ -356,16 +340,26 @@ the end points of the E edges, then 4\*F vertex indices defining F
...
+the complex face into multiple simple faces, each of which is a
+triangle or quadrilateral.
+
+.. note::
+
+  If a face is a quadrilateral then its 4 vertices must be co-planar.
+  LAMMPS does not check that this is the case.  If you have a quad-face
+  of a polyhedron that is not planar (e.g. a cube whose vertices have
+  been randomly displaced), then you should represent the single quad
+  face as two triangle faces instead.
```
Comment (ndtrung81) [inline doc/src/Howto_body.rst:363]: this is clear

**Comment by akohlmey** [comment]
@sjplimp this now tests correctly for all my test cases for all body styles and is valgrind clean (using the suppressions in tools/valgrind for known false positives). I will be merging this as soon as the integration tests clear, so I can work on other pending PRs with potentially overlapping changes and get all peripheral fixes and updates included in this PR merged as well.

**Pull request merged**